File: cpp_code/line/stack/ArrayStack.cpp
/**
* 抖码算法，让算法学习变的简单有趣
* @作者 : 老汤
*/

#include <string>
#include <iostream>
#include "Stack.h"

using namespace std;

template<typename E>
class ArrayStack: public Stack<E>{
private:
    E* data;
    int capacity;
    int size;

public:
    ArrayStack(int capacity) {
        this->capacity = capacity;
        this->data = new E[capacity];
        this->size = 0;
    }

    int getSize() {
        return size;
    }

    bool isEmpty() {
        return size == 0;
    }

    void push(E e) {
            if (size == capacity) {
                throw "push failed, Stack is full";
            }
            data[size] = e;
            size++;
        }

    E pop() {
        if (isEmpty()) {
            throw "pop failed, stack is empty";
        }
        E ret = data[size - 1];
        size--;
        return ret;
    }

    E peek() {
        if (isEmpty()) {
            throw "peek failed, stack is empty";
        }
        return data[size - 1];
    }

     string toString() {
        string res = "";
        res += "Stack: [";
        for (int i = 0; i < size; i++) {
            res += to_string(data[i]);
            if (i != size - 1) {
                res += ",";
            }
        }
         res += "] top";
        return res;
    }
};

int main() {
    ArrayStack<int> stack(10);

    stack.push(10);
    cout << stack.toString() << endl;

    stack.push(20);
    cout << stack.toString() << endl;

    stack.push(30);
    cout << stack.toString() << endl;

    stack.pop();
    cout << stack.toString() << endl;

    stack.pop();
    cout << stack.toString() << endl;

    return 0;
}

File: javascript_code/line/stack/ArrayStack.js

class ArrayStack {

    constructor(capacity) {
        this.data = new Array(capacity)
        this.size = 0
    }

    getSize() {
        return this.size
    }

    isEmpty() {
        return this.size == 0
    }

    push(e) {
        if (this.size == this.data.length) {
            throw new Error("push failed, Stack is full")
        }
        this.data[this.size] = e
        this.size++
    }

    pop() {
        if (this.isEmpty()) {
            throw new Error("pop failed, stack is empty")
        }
        const ret = this.data[this.size - 1]
        this.size--
        return ret
    }

    peek() {
        if (this.isEmpty()) {
            throw new Error("peek failed, stack is empty")
        }
        return this.data[this.size - 1]
    }

    toString() {
        let res = "Stack: ["
        for (let i = 0; i < this.size; i++) {
            res += this.data[i]
            if (i != this.size - 1) {
                res += ","
            }
        }
        res += "] top"
        return res
    }
}

const stack = new ArrayStack(5)
stack.push(10)
console.log(stack.toString())
stack.push(20)
console.log(stack.toString())
stack.push(30)
console.log(stack.toString())

stack.pop()
console.log(stack.toString())
stack.pop()
console.log(stack.toString())